test/unitTests/core/zoe/contracts/test-autoswap.js

```diff
@@ -0,0 +1,240 @@
+    const zoe = makeZoe();
+    const escrowReceiptIssuer = zoe.getEscrowReceiptIssuer();
+
+    // Setup Alice
+    const aliceMoolaPurse = mints[0].mint(issuers[0].makeAmount(10));
+    const aliceMoolaPayment = aliceMoolaPurse.withdrawAll();
+    // Let's assume that simoleans are worth 2x as much as moola
+    const aliceSimoleanPurse = mints[1].mint(issuers[1].makeAmount(5));
+    const aliceSimoleanPayment = aliceSimoleanPurse.withdrawAll();
+
+    // Setup Bob
+    const bobMoolaPurse = mints[0].mint(issuers[0].makeAmount(2));
+    const bobMoolaPayment = bobMoolaPurse.withdrawAll();
+    const bobSimoleanPurse = mints[1].mint(issuers[1].makeAmount(7));
+    const bobSimoleanPayment = bobSimoleanPurse.withdraw(
+      issuers[1].makeAmount(3),
+    );
+
+    // 1: Alice creates an autoswap instance
+
+    const { liquidityIssuer, makeAutoSwap } = makeAutoSwapMaker();
+    const allIssuers = [...issuers, liquidityIssuer];
+
+    const { zoeInstance, governingContract: autoswap } = zoe.makeInstance(
+      makeAutoSwap,
+      allIssuers,
+    );
+
+    // The issuers are defined at this step
+    t.deepEquals(zoeInstance.getIssuers(), allIssuers);
+
+    const actualLiquidityIssuer = autoswap.getLiquidityIssuer();
+    t.deepEquals(actualLiquidityIssuer, liquidityIssuer);
+
+    // 2: Alice adds liquidity
+    // 10 moola = 5 simoleans at the time of the liquidity adding
+    // aka 2 moola = 1 simolean
+    const aliceOffer = harden([
+      {
+        rule: 'haveExactly',
+        amount: allIssuers[0].makeAmount(10),
+      },
+      {
+        rule: 'haveExactly',
+        amount: allIssuers[1].makeAmount(5),
+      },
+      {
+        rule: 'wantAtLeast',
+        amount: allIssuers[2].makeAmount(10),
+      },
+    ]);
+    const alicePayments = [aliceMoolaPayment, aliceSimoleanPayment, undefined];
+
+    const {
+      escrowReceipt: allegedAliceEscrowReceipt,
+      claimPayoff: aliceClaimPayoff,
+    } = await zoeInstance.escrow(aliceOffer, alicePayments);
+
+    // 3: Alice does a claimAll on the escrowReceipt payment
+    const aliceEscrowReceipt = await escrowReceiptIssuer.claimAll(
+      allegedAliceEscrowReceipt,
+    );
+
+    const liquidityOk = await autoswap.addLiquidity(aliceEscrowReceipt);
+
+    t.equals(liquidityOk, 'Added liquidity.');
+
+    const aliceAddLiquiditySeat = await aliceClaimPayoff.unwrap();
+
+    const liquidityPayments = await aliceAddLiquiditySeat.getPayoff();
+
+    t.deepEquals(
+      liquidityPayments[2].getBalance(),
+      liquidityIssuer.makeAmount(10),
+    );
+    t.deepEquals(autoswap.getPoolQuantities(), [10, 5, 0]);
+
+    // 4: Imagine that Alice gives bob access to autoswap
+
+    // 5: Bob looks up the price of 2 moola in simoleans
+    const amount2Moola = issuers[0].makeAmount(2);
+    const simoleanAmount = autoswap.getPrice([
+      amount2Moola,
+      undefined,
+      undefined,
+    ]);
+    t.deepEquals(simoleanAmount, issuers[1].makeAmount(1));
```

Comment:
This is just an aside, since in an earlier conversation I thought you were saying that the price one gets is as quoted. Any algorithm that adjusts the price as players buy (and maintains liquidity) won't actually sell more than a marginal number of shares at its current price before the price changes.

Bob only gets quoted a price of 2 because the autoswap is losing due to roundoff. the pool goes from 10,5 to 12,4 (also throwing off the 'constant' product for the next trade). If the values were all multiplied by 100, then an offer of 200 would get 84 rather than 100.

Reply:
I think there's probably some communication problem here - a difference of terms or something. But, a user *will* get the tokens they were quoted in `getPrice` as long as no changes to the liquidity pool happened in between.